lib/docker-delta.coffee

```diff
@@ -112,16 +110,27 @@ nullDisposer = ->
 
 hardlinkCopy = (srcRoot, dstRoot, linkDests) ->
 	rsyncArgs = [
-		'--timeout', "#{RSYNC_TIMEOUT}"
 		'--archive'
 		'--delete'
 	]
 	rsyncArgs.push('--link-dest', dest) for dest in linkDests
 	rsyncArgs.push(srcRoot, dstRoot)
 	rsync = spawn('rsync', rsyncArgs)
-	utils.waitPidAsync(rsync)
+	rsync.waitAsync()
 
-exports.applyDelta = (srcImage) ->
+applyBatch = (rsync, batch, timeout) ->
+	new Promise (resolve, reject) ->
+		batch.pipe(rsync.stdin).on 'unpipe', ->
+			# wait for clean exit
+			p = rsync.waitAsync()
```

Comment:
It seems like this would be better as:
```coffee
new Promise (resolve) ->
	batch.pipe(rsync.stdin).on('unpipe', resolve)
.then ->
	rsync.waitAsync()
.then ->
	if timeout > 0
		...
```
so that the `new Promise` constructor is only around the callback code, and then we instantly go back to a normal promise chain